domain-knowledge/DOMAIN-OBJECTS.md

```diff
@@ -0,0 +1,501 @@
+- An `Artist` may force-delete their own `Album`, provided that no `Song` thereon
+  has ever been purchased. If at least one `Song` thereon has been purchased,
+  the `Artist` may soft-delete their own `Album`. If force-deleted, all associated
+  `FlacFile` records, and their binary filesystem data, are also hard-deleted. In
+  the case of soft-deletion, any `FlacFile` associated with a `Song` that has ever
+  been purchased is preserved, and all others, along with their binary filesystem
+  data, are hard-deleted. This policy ensures that customers who have purchased
+  a `Song` are able to access it indefinitely.
+
+#### Logic for Applying Full-Album Price Discount
+
+The discount amount equals the difference between the full-album price and the
+sum of the individual `Songs` on the `Album`.
+
+A simple example is an `Album` that contains 10 `Songs`, each of which is
+priced at $1.00. If the `Artist` specifies a full-album price of $8.00, then the
+discount amount equals $2.00.
+
+When the customer adds the full album to the shopping cart, the $2.00 discount
+is applied evenly over the 10 `Songs` in the shopping cart. As such, each `Song`
+would be discounted in the amount of $2.00/10 = $0.20. However, this simple example
+does not account for the potential for a fraction to result. For this reason, it
+is necessary to employ a more complicated discounting strategy. This implementation
+employs the "Distributing Percentage Off Discounts" strategy as described in
+Microsoft's Commerce Server documentation:
+
+https://docs.microsoft.com/en-us/previous-versions/commerce-server/ee785199(v=cs.20)?redirectedfrom=MSDN 
+
+This discounting strategy is not specific to Microsoft Commerce Server; it is a
+widely understood technique, but Microsoft's explanation is clear, concise,
+and includes detailed examples.
+
+### Artist
+
+An `Artist` is an entity (the term "person[s]" is avoided here because it's
+needlessly specific) who asserts authorship over an `Album`. An `Artist's` name
+is essentially a "moniker"; it does not have to be a "real name", and it could
+even be a phrase, such as _The Artist Formerly Known as Prince_, to cite a 
+familiar example.
+
+An `Artist` cannot exist without an associated `CatalogEntity`.
+
+#### Relationships
+
+- `Artist` `morphOne()` `Catalogable`
+- `Artist` `morphOne()` `Profile`
+- `Artist` `belongsTo()` `Label`
+- `Artist` `hasMany()` `Song` through `Album`
+- `Artist` `hasMany()` `Album`
+- `Artist` `morphMany()` `Featured`
+
+#### Access Policy
+
+- A `Label` has the same access as any `Artist` under its control.
+- Anybody may create an `Artist`, but it is effectively disabled until approved
+  (see `CatalogEntity` access policy).
+- The `User` associated with an `Artist's` owning `CatalogEntity` may set the
+  `label_id` field on the `Artist` object by way of accepting a `Label's`
+  invitation to join the `Label`.
+
+### CatalogEntity
+
+In essence, a `CatalogEntity` is a person or company
```

Comment:
What if `CatalogEntity` was explained before `Artist`. For me that makes the "ownership" and the role this model plays a bit easier to understand. Not a big issue though. Just a thought.

Reply:
That's an excellent point. The models are listed in alphabetical order, but maybe that's not particularly meaningful. Given that the headings are linked in the navigation, I think it makes sense to change the order, per your suggestion.

Reply:
After reading after I wrote I stand by my comment. If the CatalogEntity is moved above Aritst it makes more sense. It is already for good reason above Label and what not :)

Reply:
I concur. In fact, I re-ordered _all_ of the sections to better reflect the "flow" among their relationships.